<launch>
    <!-- GPS Navigation State -->
    <!-- This launch file defines the nodes active during GPS navigation. -->

    <!-- Our Goal-to-Velocity PID Controller -->
    <!-- <node pkg="uav_autonomous_utils" type="control/goal_to_velocity_pid.py" name="goal_to_velocity_pid" output="screen">
        <rosparam param="kp">[0.6, 0.6, 0.6, 0.8]</rosparam>
        <rosparam param="ki">[0.08, 0.08, 0.08, 0.0]</rosparam>
        <rosparam param="kd">[0.1, 0.1, 0.1, 0.1]</rosparam>
        <param name="max_speed" value="1.5" />
        <param name="max_yaw_rate" value="2.5" />
        <param name="goal_tolerance" value="0.1" />
        <param name="enable_yaw_alignment" value="true" />
        <param name="min_goal_change" value="0.2" />
        <remap from="/move_base_simple/goal" to="/move_base_mission"/>
    </node> -->

    <node pkg="uav_autonomous_utils" type="control/slidingModeController.py" name="pid_gps" output="screen">
        <rosparam param="kp">[0.5, 0.5, 0.5, 0.8]</rosparam>
        <rosparam param="ki">[0.01, 0.01, 0.01, 0.0]</rosparam>
        <rosparam param="kd">[0.1, 0.1, 0.1, 0.1]</rosparam>
        <param name="max_speed" value="2.5" />
        <param name="max_yaw_rate" value="2.0" />
        <param name="yaw_switch_distance" value="1.0" />
        <param name="position_tolerance" value="0.25" />
        <param name="yaw_tolerance" value="0.1" />
        <param name="is_gps_navigation_mode" value="true" />
        <remap from="/move_base_simple/goal" to="/move_base_mission"/>
        <remap from="/uav/odometry" to="/mavros/local_position/odom"/>
        <remap from="/uav/attractive_velocity" to="/uav/attractive_velocity"/>
    </node>

    <!-- Artificial Potential Fields Local Planner -->
    <param name="potential_fields/UAV_radius" value="0.1" />
    <node name="apf_node_gps" pkg="artificial_potential_fields" type="apf" output="screen">
        <param name="k_repulsive" value="0.02" />
        <remap from="/uav/odometry" to="/mavros/local_position/odom"/>
        <remap from="/uav/attractive_velocity" to="/uav/attractive_velocity"/>
        <remap from="/laser/scan" to="/scan_from_depth_camera"/>
        <remap from="/potential_fields/velocity" to="/mavros/setpoint_velocity/cmd_vel"/>
    </node>

    <!-- Depth Image to Laser Scan Converter (dependency for APF) -->
    <node pkg="depthimage_to_laserscan" type="depthimage_to_laserscan" name="depthimage_to_laserscan_node_gps">
        <remap from="image" to="/iris/camera/depth/image_raw"/>
        <remap from="camera_info" to="/iris/camera/depth/camera_info"/>
        <remap from="scan" to="/scan_from_depth_camera"/>
        <param name="output_frame_id" value="iris/camera_link"/>
    </node>

    <!-- VFH3D LOCAL PLANNER (LINEAR VELOCITY CALCULATOR) -->
    <!-- This node calculates the obstacle-free linear velocity and publishes it for the master controller. -->
    <!-- <node name="vfh3d_node" pkg="vfh3d" type="vfh3d_node" output="screen">
        <param name="octomap_topic" value="/octomap_binary"/>
        <param name="hist_resolution" value="0.087222222" />
        <param name="map_resolution" value="0.5" />
        <param name="max_plan_range" value="5.0" />
        <param name="const_a" value="1.0" />
        <param name="pose_topic" value="/mavros/local_position/pose" />
        <param name="cmd_pose_topic" value="/move_base_mission" />
        <param name="vehicle_safety_radius" value="0.5" />
        <param name="turning_radius_l" value="0.2" />
        <param name="turning_radius_r" value="0.2" />
        <remap from="/vfh3d/cmd_pose" to="/vfh3d/cmd_pose"/>
        <param name="vehicle_size_x" value="0.5" />
        <param name="vehicle_size_y" value="0.5" />
        <param name="vehicle_size_z" value="0.5" />
    </node> -->


    <!-- <node pkg="apf_local_planner" type="apf_local_planner" name="apf_local_planner_node" output="screen">
        <remap from="/lidar/livox" to="/iris/camera/depth/points"/>
        <remap from="/planner/goal" to="/move_base_visual"/>
        <param name="manager/max_vel" value="1.0"/>
        <param name="grid_map/obstacles_inflation" value="0.05" type="double"/>
        <param name="apf/sensor_max_range" value="5.0" type="double"/>
        <param name="apf/k_push" value="1.0" type="double"/>
        <param name="apf/k_push_xy" value="17.0" type="double"/>
        <param name="apf/k_push_z" value="0.0" type="double"/>
        <param name="apf/k_att" value="1.0" type="double"/>
        <param name="apf/min_dist" value="0.45" type="double"/>
        <param name="apf/max_att_dist" value="3.0" type="double"/>
        <param name="grid_map/ground_height" value="0.1" type="double"/>
        <param name="apf/ground_safe_height" value="0.3" type="double"/>
        <param name="apf/safe_distance" value="0.3" type="double"/>
    </node> -->


    <!-- <node name="trajectory_replanning_example" pkg="ewok_simulation" type="trajectory_replanning_example" output="screen">
      <param name="dt" value="0.5" />
      <param name="num_opt_points" value="7" />

      <param name="max_velocity" value="2.0"/>
      <param name="max_acceleration" value="5.0"/>
      <param name="resolution" value="0.15"/>
      <param name="truncation_distance" value="3.0"/>
      <param name="distance_threshold" value="0.5"/>

      <remap from="camera/depth/image_raw" to="/iris/camera/depth/image_raw"/>
      <remap from="command/point" to="/ewok_cmd_vel"/>
    </node> -->
    <!-- <node pkg="tf" type="static_transform_publisher" name="odom_to_world_tf_publisher" args="0 0 0 0 0 0 odom world 100" /> -->
























    <node pkg="uav_autonomous_utils" type="control/trajectory_follower_with_hybrid_yaw_control.py" name="trajectory_follower_with_yaw_control" output="screen">
        <!-- PID gains for Yaw -->
        <param name="kp_yaw" value="0.8" />
        <param name="ki_yaw" value="0.1" />
        <param name="kd_yaw" value="0.2" />
        <!-- Controller parameters -->
        <param name="max_yaw_rate" value="2.0" />
        <param name="yaw_switch_distance" value="1.5" />
        <param name="align_with_velocity_distance" value="10.0" />
        <param name="position_tolerance" value="0.2" />
        <param name="yaw_tolerance" value="0.1" />
        <param name="is_gps_navigation_mode" value="true" />
        <remap from="/uav/odometry" to="/mavros/local_position/odom"/>
        <remap from="/goal_point" to="/move_base_mission"/>
        <remap from="/pva_setpoint" to="/pva_setpoint"/>
        <!-- <remap from="/mavros/setpoint_raw/local" to="/mavros/setpoint_raw/local_not"/> -->
    </node>

    <!-- RAST Planning Node -->
    <!-- This node takes the risk map, performs an A* search, and optimizes the trajectory. -->
    <node pkg="rast_corridor_planning" name="trajectory_planning_node" type="planning_node" output="log">
        <rosparam file="$(find uav_autonomous_utils)/config/rast_planning.yaml" command="load" />
        <remap from="/goal_point" to="/move_base_mission"/>
    </node>






    <node pkg="depth_cbf_planner" type="cbf_local_planner.py" name="depth_cbf_local_planner" output="screen">
        <param name="odom_frame" value="odom"/>
        <param name="camera_frame" value="camera_depth_optical_frame"/>
        <param name="max_speed" value="2.5"/>
        <param name="max_yaw_rate" value="2.0"/>
        <param name="a_max" value="3.0"/>
        <param name="yaw_switch_distance" value="1.5"/>
        <param name="enable_fov" value="true"/>
        <param name="fov_half_angle_deg" value="42.0"/>
        <param name="support_k" value="1"/>
        <param name="region_radius" value="0.20"/>
        <param name="min_neighbors" value="12"/>
        <param name="dt_hold_default" value="0.02"/>
        <param name="path_rate_hz" value="10.0"/>
    </node>



</launch>